scylla-cql/src/types/deserialize/value.rs

```diff
@@ -0,0 +1,1515 @@
+                nanoseconds,
+            )));
+        }
+
+        Ok(Time(chrono::Duration::nanoseconds(nanoseconds)))
+    }
+);
+
+// inet
+
+impl_strict_type!(
+    "inet",
+    IpAddr,
+    ColumnType::Inet,
+    |_typ: &'frame ColumnType, v: Option<FrameSlice<'frame>>| {
+        let val = ensure_not_null(v)?;
+        if let Ok(ipv4) = <[u8; 4]>::try_from(val) {
+            Ok(IpAddr::from(ipv4))
+        } else if let Ok(ipv16) = <[u8; 16]>::try_from(val) {
+            Ok(IpAddr::from(ipv16))
+        } else {
+            Err(ParseError::BadIncomingData(format!(
+                "Invalid inet bytes length: {}",
+                val.len(),
+            )))
+        }
+    }
+);
+
+// uuid
+// TODO: Consider having separate types for timeuuid and uuid
+
+impl_strict_type!(
+    "timeuuid or uuid",
+    Uuid,
+    ColumnType::Uuid | ColumnType::Timeuuid,
+    |_typ: &'frame ColumnType, v: Option<FrameSlice<'frame>>| {
+        let val = ensure_not_null(v)?;
+        let arr = ensure_exact_length::<16>("timeuuid or uuid", val)?;
+        let i = u128::from_be_bytes(arr);
+        Ok(uuid::Uuid::from_u128(i))
+    }
+);
+
+/// A value that may be empty or not.
+///
+/// In CQL, some types can have a special value of "empty", represented as
+/// a serialized value of length 0. An example of this are integral types:
+/// the "int" type can actually hold 2^32 + 1 possible values because of this
+/// quirk. Note that this is distinct from being NULL.
+///
+/// `MaybeEmpty` was introduced to help support this quirk for Rust types
+/// which can't represent the empty, additional value.
+pub enum MaybeEmpty<T> {
+    Empty,
+    Value(T),
+}
+
+impl<'frame, T> DeserializeCql<'frame> for MaybeEmpty<T>
+where
+    T: DeserializeCql<'frame>,
+{
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+        <T as DeserializeCql<'frame>>::type_check(typ)
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        let val = ensure_not_null(v)?;
+        if val.is_empty() {
+            Ok(MaybeEmpty::Empty)
+        } else {
+            let v = <T as DeserializeCql<'frame>>::deserialize(typ, v)?;
+            Ok(MaybeEmpty::Value(v))
+        }
+    }
+}
+
+// secrecy
+#[cfg(feature = "secret")]
+impl<'frame, T> DeserializeCql<'frame> for secrecy::Secret<T>
+where
+    T: DeserializeCql<'frame> + secrecy::Zeroize,
+{
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+        <T as DeserializeCql<'frame>>::type_check(typ)
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        <T as DeserializeCql<'frame>>::deserialize(typ, v).map(secrecy::Secret::new)
+    }
+}
+
+// collections
+
+// lists and sets
+
+/// An iterator over either a CQL set or list.
+pub struct SequenceIterator<'frame, T> {
```

Comment:
I'm not sure I like the term "listlike". It leans too heavily towards "list" and not enough towards "set".

On the other hand, "sequence" is not perfect either as it implies that the elements are ordered (which is not true in case of a set).